src/lib/reasoners/intervalCalculus.ml

```diff
@@ -2035,6 +2041,79 @@ let case_split env uf ~for_model =
       end
   | _ -> res
 
+let optimizing_split env uf opt_split =
+  (* soundness: if there are expressions to optmize, this should be
+     done without waiting for ~for_model flag to be true *)
+  let {Th_util.order; r = r; is_max = to_max; e; value } = opt_split in
+  assert (match value with
+      | Unknown -> true
+      | _ -> false
+    );
+  let repr, _ = Uf.find uf e in
+  let ty = E.type_info e in
+  let r1 = r in (* instead of repr, which may be a constant *)
+  let p = poly_of repr in
+  match P.is_const p with
+  | Some optim ->
+    Printer.print_dbg "%a has a %s: %a@."
+      E.print e
+      (if to_max then "maximum" else "minimum")
+      Q.print optim;
+
+    let r2 = alien_of (P.create [] optim  ty) in
+    Debug.case_split r1 r2;
+    let t2 = mk_const_term optim ty in
+    let o =
+      Some {Th_util.opt_ord = order; opt_val = Th_util.Value t2}
+    in
+    let s =
+      LR.mkv_eq r1 r2, true, Th_util.CS (o, Th_util.Th_arith, Q.one)
+    in
+    Some { opt_split with value = Value s }
+
+  | None ->
+    begin
+      let sim = if ty == Ty.Tint then env.int_sim else env.rat_sim in
+      let p = if to_max then p else P.mult_const Q.m_one p in
+      let l, c = P.to_list p in
+      let l = List.rev_map (fun (x, y) -> y, x) (List.rev l) in
+      let sim, mx_res = Sim.Solve.maximize sim (Sim.Core.P.from_list l) in
+      match Sim.Result.get mx_res sim with
+      | Sim.Core.Unknown -> assert false
+      | Sim.Core.Sat _   -> assert false (* because we maximized *)
+      | Sim.Core.Unsat _ -> assert false (* we know sim is SAT *)
+      | Sim.Core.Unbounded _ ->
+        Printer.print_dbg
+          "%a is unbounded. Let other methods assign a value for it@."
+          E.print e;
+        let value =
+          if to_max then Th_util.Pinfinity else Th_util.Minfinity
+        in
+        Some { opt_split with value }
+
+      | Sim.Core.Max (lazy Sim.Core.{ max_v; is_le = true }, _sol) ->
+        let max_p = Q.add max_v.bvalue.v c in
+        let optim = if to_max then max_p else Q.mult Q.m_one max_p in
+        Printer.print_dbg "%a has a %s: %a@."
+          E.print e
+          (if to_max then "maximum" else "minimum")
+          Q.print optim;
+
+        let r2 = alien_of (P.create [] optim  ty) in
+        Debug.case_split r1 r2;
+        let t2 = mk_const_term optim ty in
+        let o =
+          Some {Th_util.opt_ord = order; opt_val = Th_util.Value t2}
+        in
+        let s =
+          LR.mkv_eq r1 r2, true, Th_util.CS (o, Th_util.Th_arith, Q.one)
+        in
+        Some { opt_split with value = Value s; }
+
+      | Sim.Core.Max (lazy Sim.Core.{ is_le = false; _ }, _) ->
+        (* There is no upper bound as we try to optimize a strict bound. *)
+        None
```

Comment:
This is not true for integers, I don't know if ocplib-simplex can return `is_le = false` for an integer problem here but we need to check (and probably assert `ty == Treal` here).

Reply:
Also I don't think returning `None` is good here — `None` means "I don't know how to optimize this expression", but we *do* know how to optimize this expression! We should have an additional constructor `StrictBound` in `optimized_split_value` to keep track of this instead.